testing.md

```diff
@@ -34,6 +34,82 @@ A piece of code is not done until it's tested.
+event when the UI is finished refreshing, and make the test listen for this event.
+
+Testing non-deterministic code can also be very difficult. If you can, avoid non-deterministic
+behavior in your code. If you must have it, try to isolate it so it can be deterministically mocked
+in tests.
+
+One type of non-determinism that *is* useful for testing is generating random inputs.
+[Generative testing](http://blog.8thlight.com/connor-mendenhall/2013/10/31/check-your-work.html), in
+which a program generates different test inputs each time the tests are run, can help describe
+properties of your program that should hold for all inputs. They can make a stronger assertion about
+correctness, since, if they pass, it means your program works for many random inputs, not just a few
+representative ones. Generative tests can also come up with edge cases that you may not have thought
+of.
+
+### Prefer Readability and Ease of Code
+
+Tests should be written in a manner that prioritizes readability and ease of writing new tests and
+modifying existing ones. In particular, these should be prioritized over other typical concerns like
+abstraction and removing code duplication. If abstractions are needed, they should be created in a
+manner that preserves the logic in the test functions code without requiring many hops. Another way
+of saying this is that it should be apparent what the test is doing by just looking at the test
+code.
+
+As an example, we had a testing library called which populated a bunch of fake data in the database
+specified by the test. A lot of our tests used with a single statement of the form
+`db_test_helpers.set_up_test_data`. This single function populated a bunch of state within the
+database including creating students, teachers, districts, apps. However, it also hid away all the
+details of the state being created. This meant that it became very hard to reason about tests or
+modify your tests if the current state was not appropriate for them.
+
+Instead, a better approach is to provide easy abstractions for creating individual pieces of data
+like `add_student`, `add_district` and use them as needed in different tests.
+
+### Ensure Tests are Independent
+
+Each test should be written in a manner that makes it independent of other tests. If there is any
+state needed for a test, the test should be responsible for getting to the state before running it
+and resetting it after execution. That is, the state before and after the test should be identical
+whether the test passes or fails. If your tests are independent, then running the tests in any order
+should not affect the outcome of the tests.
+
+A stronger version of independence could mean that there is no shared state between different tests.
+As an example, if the test involves talking to a database, then different tests should talk to
+different logical databases. Another example, if you are testing the behavior of a particular class,
```

Comment:
setting up a different db for each test case seems costly in terms of time. that's the only real advantage to shared state that i can think of. how do you balance that concern?